### DISCUSSION TITLE
API : admin/drive/cleanup ?

### DISCUSSION THREAD

**Discussion by kazukyakayashi** [misskey-dev/misskey#7304] [Q&A]
What exactly does this action do ? (compared to the other two that are : `admin/drive/clean-remote-files ` and `admin/delete-all-files-of-a-user `)

[Misskey API](https://misskey.io/api-doc#operation/admin/drive/cleanup)

**Comment by Johann150** [comment]
⚠️ **Reviewing this now, I would not recommend using this API endpoint. Please see #8222.** ⚠️

***

Looking at the [code for this endpoint](https://github.com/misskey-dev/misskey/blob/develop/packages/backend/src/server/api/endpoints/admin/drive/cleanup.ts) it deletes drive files whose `userId` (i.e. "owner") is `null`. In combination with this database constraint
```sql
"FK_860fa6f6c7df5bb887249fba22e" FOREIGN KEY ("userId") REFERENCES "user"(id) ON DELETE SET NULL
```
specifically the `ON DELETE SET NULL` part, this looks like it was intended to delete orphaned drive files (i.e. drive files that are owned by deleted users).
But it is currently problematic because `userId` being `null` can also indicate that the owner is the "system", which is for example the case for custom emoji copied from other instances. See the issue linked above.

**Reply by Johann150** [reply]
For completeness:
- `admin/drive/clean-remote-files` deletes all files which do not belong to local users (i.e. have been cached from remote instances).
- `admin/delete-all-iles-of-a-user` deletes all files associated with a specific user (could be a local or remote user).
